### PR TITLE
Use system brotli on Unix non-portable builds

### LINKED ISSUE #107020: source-built NativeAOT doesn't work with system brotli
When source-building .NET with the brotli system lib, the source-built NativeAOT fails to compile an application that uses brotli.

The compilation failure is due to missing flags for brotli.

```
  console failed with 2 error(s) (12.2s) → bin/Release/net9.0/fedora.40-x64/console.dll
    clang : error : linker command failed with exit code 1 (use -v to see invocation)
    /var/home/tmds/.nuget/packages/microsoft.dotnet.ilcompiler/9.0.0-rc.2.24423.10/build/Microsoft.NETCore.Native.targets(376,5): error MSB3073: The command ""clang" "obj/Release/net9.0/fedora.40-x64/native/console.o" -o "bin/Release/net9.0/fedora.40-x64/native/console" -Wl,--version-script=obj/Release/net9.0/fedora.40-x64/native/console.exports -Wl,--export-dynamic -gz=zlib -fuse-ld=bfd /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/sdk/libbootstrapper.o /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/sdk/libRuntime.WorkstationGC.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/sdk/libeventpipe-disabled.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/sdk/libRuntime.VxsortEnabled.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/sdk/libstandalonegc-disabled.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/sdk/libstdc++compat.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/framework/libSystem.Native.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/framework/libSystem.Globalization.Native.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/framework/libSystem.IO.Compression.Native.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/framework/libSystem.Net.Security.Native.a /tmp/dotnet/packs/runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler/9.0.0-rc.2.24423.10/framework/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -pthread -lssl -lcrypto -ldl -lz -lrt -lm -pie -Wl,-pie -Wl,-z,relro -Wl,-z,now -Wl,--eh-frame-hdr -Wl,--discard-all -Wl,--gc-sections" exited with code 1.
```

In https://github.com/dotnet/runtime/pull/97536, we've added such flags for OpenSSL:
https://github.com/dotnet/runtime/blob/9a31a5b47aa7484bec71ba28aa59712c7898b902/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L166-L169

We need to additionally have:
```xml
      <NativeSystemLibrary Include="brotlienc" />
      <NativeSystemLibrary Include="brotlidec" />
```

.NET can be either built with system brotli or the bundled broti. The above is only needed when built with system brotli.

`Microsoft.NETCore.Native.Unix.targets` is in the `Microsoft.DotNet.ILCompiler` package, which means it may come from nuget.org, so we can't just patch it as part of source-building .NET.

For the OpenSSL flags, https://github.com/dotnet/runtime/pull/97536 introduced a `nonportable.txt` breadcrumb that is stored in the source-built .NET so `Microsoft.DotNet.ILCompiler` knows when to include the OpenSSL flags.

For brotli, we need something similar.

If we can consider brotli an acceptable system dependency for source-build platforms, we may consider to have non-portable builds always use brotli and use the `nonportable.txt` breadcrumb.

cc @jkotas @omajid @MichaelSimons

### PR DESCRIPTION
**Pull request by agocke**
Fixes #107020

### CODE REVIEW

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
Comment (jkotas) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:168]:
```suggestion
      <NativeSystemLibrary Include="brotlienc" />
      <NativeSystemLibrary Include="brotlidec" />
```
brotlicommon is only used for static linking: https://github.com/dotnet/runtime/pull/107166#discussion_r1739122983

Also, we may want to follow the style of other `NativeSystemLibrary` items.

**Comment by jkotas** [comment]
@tmds Could you please validate that this is fixing #107020?

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
Comment (tmds) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: Can we extend this with something that is similar to https://github.com/dotnet/runtime/pull/107166/files#r1739122333? Then source-build can be built with bundled or system brotli.

Otherwise we need to add `--cmakeargs -DCLR_CMAKE_USE_SYSTEM_BROTLI=true` (similar to what is in the diff in https://github.com/dotnet/runtime/issues/107020#issuecomment-2320204908) and source-building requires system brotli.

**Comment by tmds** [comment]
> @tmds Could you please validate that this is fixing https://github.com/dotnet/runtime/issues/107020?

Yes, I will do some builds to validate this.

**Comment by tmds** [comment]
cc @omajid @MichaelSimons

**Comment by tmds** [comment]
> Yes, I will do some builds to validate this.

https://github.com/dotnet/runtime/pull/107225#discussion_r1739627515 needs to be tackled first.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: For .NET 10, we will get it with #107166.

For .NET 9, it would require build refactoring that is hard to justify for a backport at this point.

Comment (tmds) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: ok, then we need to add something that sets `CLR_CMAKE_USE_SYSTEM_BROTLI=true` so maintainers provide brotli in their non-portable builds.

Without it, the brotli the bootstrap build will work (since Microsoft's portable SDK doesn't have a `nonportable.txt` breadcrumb) but successive builds using the produced non-portable SDK won't, because NativeAOT will fail when trying to find brotli on a system that doesn't provide it.

Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: Why is this not sufficient https://github.com/dotnet/runtime/blob/main/eng/DotNetBuild.props#L91 ?

Comment (tmds) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: That doesn't do anything unless the maintainer builds the vmr with `--with-system-libs brotli`.

If he doesn't add it, his bootstrap build works, but he is not aware that successive builds using that SDK will fail unless he provides brotli.

By adding `CLR_CMAKE_USE_SYSTEM_BROTLI=true` for non-portable, we make this issue visible also on the bootstrap build.

Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: > That doesn't do anything unless the maintainer builds the vmr with --with-system-libs brotli.

Are there any distros that do not want to use `--with-system-libs brotli`? I think that the conversation in https://github.com/dotnet/runtime/issues/107020 suggested that all interested distro builds want to use `--with-system-libs brotli`.

Comment (tmds) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: Yes, there are no maintainers that have objected to using system brotli, and we are assuming brotli to be a common available system library.

Note that this change will also require a brotli development package to be added to the source-build CI jobs.

Comment (agocke) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: By that you mean added to the container?

Comment (tmds) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:167]: Yes.

**Comment by tmds** [comment]
> Have you validated that this fix works locally? I can start upgrading the containers separately.

Can we address https://github.com/dotnet/runtime/pull/107225#discussion_r1739627515 first?

I'll validate this by building the vmr, and then building it again using the output of the first build.
It's going to take a few hours to build, so it would be preferable to validate the complete patch.

**Comment by tmds** [comment]
> Can we address https://github.com/dotnet/runtime/pull/107225#discussion_r1739627515 first?

By this, I mean: add something to `DotNetBuild.props` similar to what is shown in this diff: https://github.com/dotnet/runtime/issues/107020#issuecomment-2320204908.

**Comment by tmds** [comment]
@agocke I've validated this change, and it is working as expected both for the portable publish as the non-portable publish.

**Comment by agocke** [comment]
It looks like this works now, after the update to the containers. From the build log:

>     Executing "/__w/1/s/artifacts/sb/src/src/coreclr/build-runtime.sh" -cmakeargs " -DCLR_CMAKE_USE_SYSTEM_BROTLI=true" -x64 -release -ci -portablebuild=false -keepnativesymbols -os linux -outputrid linux-x64 -cmakeargs "-DCLR_DOTNET_HOST_PATH=/__w/1/s/.dotnet/dotnet" -cmakeargs "-DCDAC_BUILD_TOOL_BINARY_PATH=/__w/1/s/artifacts/sb/src/artifacts/bin/coreclr/linux.x64.Release/cdac-build-tool/cdac-build-tool.dll"

**Comment by agocke** [comment]
@tmds This is now waiting on your review, let me know if there's anything else needed to unblock the build.

**Review by tmds — APPROVED** [review]
Lgtm! Thanks!

**Comment by agocke** [comment]
@jkotas Anything else I should cover?

**Review by jkotas — APPROVED** [review]
LGTM

**Comment by jkotas** [comment]
@jkoritzinsky This change should be reverted in your PR

**Comment by am11** [comment]
This is failing Android build in SDK codeflow: https://dev.azure.com/dnceng-public/public/_build/results?buildId=818433&view=logs&jobId=66673ce3-9c46-54b2-cf79-862a2d55b22a&j=66673ce3-9c46-54b2-cf79-862a2d55b22a&t=f73a763b-d161-54c1-b189-48197f51b0b6

https://github.com/dotnet/sdk/pull/43070

```
      Running '   cmake -DCMAKE_INSTALL_PREFIX="/vmr/src/runtime/artifacts/obj/mono/android.arm64.Release/out" -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release  -DCLR_CMAKE_USE_SYSTEM_BROTLI=true -DGC_SUSPEND=hybrid -DMONO_LIB_NAME=monosgen-2.0 -DMONO_SHARED_LIB_NAME=monosgen-2.0 -DCMAKE_TOOLCHAIN_FILE=/usr/local/android-ndk/build/cmake/android.toolchain.cmake -DANDROID_NDK=/usr/local/android-ndk -DANDROID_STL=none -DANDROID_CPP_FEATURES="no-rtti no-exceptions" -DANDROID_PLATFORM=android-21 -DANDROID_ABI=arm64-v8a -DENABLE_MINIMAL=ssa,logging -DFEATURE_PERFTRACING_PAL_TCP=1 -DFEATURE_PERFTRACING_DISABLE_DEFAULT_LISTEN_PORT=1 -DMONO_COMPONENTS_RID=android-arm64 -DCLR_CMAKE_HOST_ARCH=arm64 -DCMAKE_C_FLAGS=" -fpic -fstack-protector -DANDROID64 -DL_cuserid=9 -D__POSIX_VISIBLE=201002 -DSK_RELEASE -DNDEBUG -UDEBUG -Wl,--build-id=sha1" -DCMAKE_CXX_FLAGS=" -fpic -fstack-protector -DANDROID64 -DL_cuserid=9 -D__POSIX_VISIBLE=201002 -DSK_RELEASE -DNDEBUG -UDEBUG -Wl,--build-id=sha1"  "/vmr/src/runtime/src/mono"' in '/vmr/src/runtime/artifacts/obj/mono/android.arm64.Release/'
      -- The C compiler identification is Clang 12.0.9
      -- The CXX compiler identification is Clang 12.0.9
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Check for working C compiler: /usr/local/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang - skipped
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: /usr/local/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- The linker identification is LLD 12.0.9 (/buildbot/src/android/llvm-r416183/out/llvm-project/lld c935d99d7cf2016289302412d708641d52d2f7ee) (compatible with GNU linkers)
      CMAKE_SYSTEM_NAME=Android
      CMAKE_SYSTEM_VARIANT=
      CMAKE_SYSTEM_PROCESSOR=arm64
      TARGET_ARCH=arm64
      CMAKE_CROSSCOMPILING=TRUE
      CMake Error at /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/extra_libs.cmake:24 (find_library):
        Could not find BROTLIDEC using the following names: brotlidec
      Call Stack (most recent call first):
        CMakeLists.txt:517 (append_extra_compression_libs)
```
cc @lewing

**Comment by am11** [comment]
Fix: https://github.com/dotnet/arcade/pull/15107. After merging, we will need to trigger https://github.com/dotnet/dotnet-buildtools-prereqs-docker/tree/main/src/azurelinux/3.0/net9.0/android build. cc @mthalman

**Comment by am11** [comment]
Thanks! `"created": "2024-09-26T17:13:41.4665931Z",` (20 hrs. ago) and the job that ran afterwards (15 hrs. ago) is [passing](https://dev.azure.com/dnceng-public/public/_build/results?buildId=819786&view=logs&j=66673ce3-9c46-54b2-cf79-862a2d55b22a), while the previous one [failed](https://dev.azure.com/dnceng-public/public/_build/results?buildId=818433&view=logs&j=7fd527c7-3836-51a9-dabe-475f4f60bb5d). 👍

**Pull request merged**
